### DISCUSSION TITLE
High Availability for Manger node

### DISCUSSION THREAD

**Discussion by 5MI7th3MI6** [Security-Onion-Solutions/securityonion#9068] [Unsupported Versions] [answered]
Hello, I am setting up a fully distributed environment does anyone have insight on how to set up high availability for the manger node, forward node, and search node?

**Comment by EbolaWare** [comment]
So, in my work center, High Availability has been a topic of discussion for a while. (Pretty much since SecurityOnion was decided on.)

In my understanding, HA was (and is) not really a desired development goal for the **manager** node. Granted, I'm not privy to the inner workings of SecOn, so that may or may not be the case. **Receiver** nodes are useful for offloading the logstash and redis portion of the **manager**.

**Search** nodes are easily expanded, simply by installing more. 

As for the **Forward** nodes, I am not sure what you would want HA with. I suppose you could have an A and B node, with a load balancer between your tap or span and the Forward nodes, but that seems outside the scope of what SO is intended to deliver.

For the other critical functions of the manager (specifically the _lone_ Elasticsearch master node and the salt-master) I'm not aware of any way within the current bounds of the project to expand those. I am working on pasting together a new type of node to at least handle the salt-master function, but my time is frequently absorbed in other projects.

**Comment by TOoSmOotH** [comment, marked as answer]
You can achieve HA on all parts of the infrastructure sans the web interface and the salt master in 2.3. You can modify the minion pillar of the ES nodes to add additional elastic roles such as "master" or "transform"  to other search nodes. You would need to enable replicas though for it to be of any use from an HA standpoint assuming you are using a true cluster. Starting in 2.4 we are using true elastic clustering by default. Receiver nodes are the way to make your pipeline redundant which is the most important thing to have redundancy for. That way if you manager goes down you will continue to index events until you bring it back up. We have seen users have success with creating redundancy on the manager via using VM based tools such as vMotion etc.  

To allow the use of minion masters in salt we need to change how we do some things. Multiple masters in salt is not trivial from the standpoint of key management and keeping all the things in sync. We have made some fundamental changes to how we are handling things in 2.4 that could enable this one day.

TLDR: I would go ahead and set up an additional receiver node so that your data pipeline has redundancy.

**Reply by EbolaWare** [reply]
Just finished the not-so-trivial task of creating a second salt-master. 

I've got it built into a salt state, and a script or 2. Essentially, replace all calls for `set MANAGER = [...] 'master' ` in all the SLS files. Since they were inconsistent between pillar.get and grains.get , I set them all to grains.get since this should improve performance. On all the minions, I created a new grain `so_manager` that has the IP or DNS name of the manager. Then, in `/etc/salt/minion`, you have to update the `manager: '<so_manager>'` field to be a list.
```
... 
manager:
  - '<so_manager>'
  - '<salt-master2>'
... 
```
Then, restarting the minions should be all it takes.

I would also recommend storing this as a state, that way it can be re run after any upgrades... If necessary...
